bnxt: Add support for udp_tunnel_port_add/del dev_ops

ajitkhaparde opened the pull request:
Add support for udp_tunnel_port_add/del dev_ops to configure a UDP port
for VXLAN and Geneve Tunnel protocols.

Signed-off-by: Ajit Khaparde <ajit.khaparde@broadcom.com>

ajitkhaparde committed:
bnxt: Add support for udp_tunnel_port_add/del dev_ops

Add support for udp_tunnel_port_add/del dev_ops to configure a UDP port
for VXLAN and Geneve Tunnel protocols.

Signed-off-by: Ajit Khaparde <ajit.khaparde@broadcom.com>

ajitkhaparde committed:
bnxt: Update tx offload capabilities

Update the tx offload capabilities of the PMD.
Signed-off-by: Ajit Khaparde <ajit.khaparde@broadcom.com>

stephen-hurd commented on drivers/net/bnxt/hsi_struct_def_dpdk.h:
```diff
@@ -6560,6 +6562,188 @@ struct hwrm_cfa_l2_set_rx_mask_output {
+	uint8_t tunnel_type;
+	/* Tunnel Type. */
+	/* Virtual eXtensible Local Area Network (VXLAN) */
+	#define TUNNEL_DST_PORT_ALLOC_REQ_TUNNEL_TYPE_VXLAN       UINT32_C(0x1)
```
This field is only 8-bits wide... this should likely be UINT8_C()  Same problem in other places, I suppose it's fine for now but we need to look at this through the whole header.

stephen-hurd reviewed (approved):
Approved, though we need to have a conversation about constant sizes.

stephen-hurd commented on drivers/net/bnxt/bnxt_hwrm.c:
```diff
@@ -2027,6 +2038,50 @@ int bnxt_hwrm_pf_evb_mode(struct bnxt *bp)
+
+	switch (tunnel_type) {
+	case TUNNEL_DST_PORT_ALLOC_REQ_TUNNEL_TYPE_VXLAN:
+		bp->vxlan_fw_dst_port_id = resp->tunnel_dst_port_id;
```
This overwrites the port ID, but maintains a count.  This looks like a resource leak.

ajitkhaparde replied:
Right. We support only one port. Will add code to prevent the over-write.

stephen-hurd replied:
This still allows one of each... do we support that?

stephen-hurd reviewed (changes requested):
What is the count for if we only have one?

ajitkhaparde committed:
bnxt: Modify UDP port handling code.

Cache the configured VXLAN/Geneve ports and use that value to check
if the right UDP port is being freed up.
Since only one UDP port can be configured for each tunnel type
prevent over-writing an existing port with a new value.
Skip calling bnxt_hwrm_tunnel_dst_port_alloc if the same UDP port is
being programmed.
Skip calling bnxt_hwrm_tunnel_dst_port_free if no UDP port has been
configured.

Signed-off-by: Ajit Khaparde <ajit.khaparde@broadcom.com>

ajitkhaparde commented:
Can you check if the new logic handles the case that you pointed out. Thanks

ajitkhaparde commented:
How does this look now?

ajitkhaparde committed:
bnxt: Use ref count to manage udp port config

The HWRM supports only one global destination port for a tunnel type,
use a referene counter to keep track of its usage.

Signed-off-by: Ajit Khaparde <ajit.khaparde@broadcom.com>
